[DSCP-377] Logic and datatypes for handling and validating Fair CVs

flyingleafe opened the pull request:
### Description

This PR should contain the following:
- [x] Datatypes defining Fair CV
- [x] Pure validation logic for Fair CV
- [x] JSON instances for Fair CV (hand-written)
- [x] Witness API endpoint for Fair CV validation against the public chain

Also, this PR will possibly include a lot of rewriting of Merkle proof logic and representation.

### YT issue

https://issues.serokell.io/issue/DSCP-377

### Checklist

Hint: a perfect PR has all the checkmarks set.

Possible related changes (conditional):
- [x] I added tests if required, in case they are:
  - Covering new introduced functionality (feature).
  - Regression tests preventing the bug from silently reappearing again (bugfix).
- [x] I have checked the [sample config](../tree/master/docs/config-full-sample.yaml) and [launch scripts](../tree/master/scripts/launch) and updated them if it was necessary (especially if executables or CLIs were changed).
- [x] I have checked READMEs and changed them accordingly if it's necessary (the main [README.md](../tree/master/README.md) as well as subproject READMEs for all related executables).
- [x] If a public API structure or/and data serialization was changed, I made sure that these changes are reflected in the:
  - [Swagger](../tree/master/specs/disciplina) API specs.
  - Any [related documentation](../tree/master/docs/api-types.md).
- [x] If any public documentation was written, I have spellchecked it.

Stylistic (obligatory):
- [x] My commit history is clean, descriptive and do not contain merge or revert commits or commits like "WIP".
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).
- [x] My code is documented (haddock) and these docs are decent (full enough and spellchecked).

flyingleafe committed:
[DSCP-377] Core datatypes and pure validation for Fair CV

flyingleafe committed:
[DSCP-377] Splitting and merging Merkle proofs + tests

Martoon-00 commented on core/src/Dscp/Crypto/MerkleTree.hs:
```diff
@@ -285,15 +300,66 @@ fillEmptyMerkleTree plugs (Empty sieve) =
+        go l <*>
+        (modify' (first (+ pnSize l)) *> go r)
+    go (ProofPruned s)     = pure $ ProofPruned (coerceSig s)
+    go (ProofLeaf s a)     = pickLeaf a $> ProofPruned (coerceSig s)
```
It looks awesome.
First, you make `EmptyMerkleTree` be equal `MerkleTree Void`, and now you automatically get only two constructors of `MerkleTree` available to be returned value here.

flyingleafe replied:
@Martoon-00 actually, the whole story with merkle trees should be rewritten, because `MerkleProof`s in general should not contain inner nodes - only leaves and pruned ends; all the inner nodes should be calculated during validation. Keeping all the inner nodes is a waste of memory, since they should be recalculated during validation anyway.

Martoon-00 commented on core/tests/Test/Dscp/Crypto/MerkleTree.hs:
```diff
@@ -66,10 +83,14 @@ spec_merkleTree = describe "Merkle Tree Tests" $ do
+
+    it "splits and merges proofs and its data correctly" $ property $
+        \(proof :: MerkleProof Int) ->
+            uncurry mergeProofAndData (separateProofAndData proof) === Just proof
```
It's also interesting to see them called vice-versa - `separateProfAndData $ mergeProofAndData (proof, data)`
(unless `mergeProofAndData` also takes only what `separateProfAndData` produces).

Martoon-00 reviewed (commented):
I very like you changes.
Looking forward for the next two points in your list.

Martoon-00 commented on core/src/Dscp/Core/FairCV.hs:
```diff
@@ -0,0 +1,130 @@
+-- error.
+validateFairCV :: FairCV Unchecked -> Either Text (FairCV Valid)
+validateFairCV (FairCV cv) =
+    FairCV <$> traverse (traverse validateTaggedProof) cv
```
Btw, according to your plan, we probably want it to return the exact list of false headers.

flyingleafe committed:
[DSCP-377] Aeson instances for FairCV

flyingleafe committed:
[DSCP-377] Endpoint for FairCV validation

flyingleafe committed:
[DSCP-377] Tests for FairCV validation

flyingleafe committed:
[DSCP-377] Update Witness API specs with new endpoint

flyingleafe committed:
[DSCP-377] Make FairCV JSON representation expressible in OpenAPI 3

OpenAPI 3.0 does not support defining heterogenous arrays with fixed
length, since `items` property cannot be an array and there is no
`additionalItems` property, which is present in JSON Schema 6.0.
Therefore, pairs should be encoded as objects, not as arrays of two
elements.

flyingleafe commented:
@Martoon-00 @Heimdell not WIP anymore, please review.

Martoon-00 reviewed (changes requested):
I like your work.
Have a couple of concerns though, please take a look.
Also, comment on my previous notes.

I'll make one another run on validation logic a bit later.

flyingleafe commented on specs/disciplina/witness/api/witness.yaml:
```diff
@@ -347,18 +352,44 @@ paths:
                     $ref: '#/components/schemas/Hash'
 
+  /checkcv:
+    post:
```
@Martoon-00 Fair question. Mainly because serializing a FairCV (which might be quite big) into a GET query string seems like a shitty solution. But I agree that the choice of a method is non-RESTy. We can have a middle-ground here and use PUT method, to emphasize the idempotency of it.

Martoon-00 replied:
Right, makes sense. Please leave a comment on it somewhere in the code.
I'm fine with 'POST', since I've heard a lot of people circumventing `GET` restrictions with `POST`, but not with `PUT`.

flyingleafe replied:
@Martoon-00 why not with `PUT`? It is _semantically_ incorrect to use anything but `GET` here, strictly speaking, but using `PUT` at least signifies that the method is idempotent (and it is), while `POST` says nothing useful about the method.

Martoon-00 replied:
Well, that makes a fair point. Let's really go with `PUT`

flyingleafe commented on witness/src/Dscp/Witness/Web/Logic.hs:
```diff
@@ -201,3 +204,19 @@ getHashType someHash = fmap (fromMaybe HashIsUnknown) . runMaybeT . asum $
+    checkAgainstDB (FairCV cv) = FairCVCheckResult <$>
+        M.traverseWithKey (M.traverseWithKey . checkProofAgainstDB) cv
+    checkProofAgainstDB addr h proof =
+        maybe False (checkProofPure addr proof) <$>
```
@Martoon-00 You're right, and `FairCVCheckResult` should be way more complex eventually - it should include not only the detailed type of failure in case of failure, but also assurance level in case of the success (i. e. the depth of the block in which the publication resides) But I intend to keep it simple in this PR and elaborate on the check result separately.

flyingleafe commented on witness/src/Dscp/Witness/Logic/Getters.hs:
```diff
@@ -195,3 +196,9 @@ getPrivateTipHash
+    :: PrivateHeaderHash
+    -> SdM_ chgacc (Maybe PublicationTxWitnessed)
+getPublicationByHeaderHash = runMaybeT . fmap piTw .
+    (MaybeT . SD.queryOne >=> MaybeT . SD.queryOne)
```
Fair point, I'll throw `LEMalformed` in the latter case.

flyingleafe commented on witness/tests/Test/Dscp/Witness/Explorer/ExplorerSpec.hs:
```diff
@@ -78,158 +100,247 @@ getTransactionsPaged chunkSize mAddress = getFrom Nothing
-        tx <- createAndSubmitPub selectGenesisSecret
-        res <- lift $ getHashType (toHex $ toPtxId tx)
-        return $ res === HashIsPublicationTx
+    describe "getTransaction" $ do
```
@Martoon-00 Setting up different indentation rules for different files in the project seems like a shitty idea, because it means complicating editor settings for everyone. There are only few specs which use two-spaces indentation, and frankly, I feel ashamed that I apparently missed it during some previous reviews.

Martoon-00 replied:
Meh, okay

flyingleafe commented on witness/tests/Test/Dscp/Witness/Explorer/ExplorerSpec.hs:
```diff
@@ -78,158 +100,247 @@ getTransactionsPaged chunkSize mAddress = getFrom Nothing
+            checkResBad <- lift $ checkFairCV fairCvInvalid
+            return $
+                fullyValid checkResGood &&
+                not (fullyValid checkResBad)
```
:man_facepalming: Yeah, I just didn't remember `counterexample` when I tried to remember how to do it.

Martoon-00 replied:
Yeah, it took me a while as well, pretty unobvious naming.

flyingleafe commented on witness/tests/Test/Dscp/Witness/Explorer/ExplorerSpec.hs:
```diff
@@ -78,158 +100,247 @@ getTransactionsPaged chunkSize mAddress = getFrom Nothing
+                    M.lookup hhash subCv
+        newProof <- mkTaggedProof <$> spoilMerkleProof (unTaggedProof proof)
+        return $ FairCV $
+            cv & at addr . _Just . at hhash . _Just .~ newProof
```
Another proof of the fact that my brain is slowly decaying due to insufficient programming practice and that I'm going to die soon.

flyingleafe commented on witness/tests/Test/Dscp/Witness/Explorer/ExplorerSpec.hs:
```diff
@@ -78,158 +100,247 @@ getTransactionsPaged chunkSize mAddress = getFrom Nothing
+    deriving (Show, Eq, Generic, Enum, Bounded)
+
+-- | Spoils a @'MerkleProof'@ by changing some of its internal nodes
+-- TODO: spoil better.
```
meaning: "change not only the root node but inner nodes sometimes too".
This function will be rewritten properly (tm) anyway in scope of DSCP-396

flyingleafe commented on witness/tests/Test/Dscp/Witness/Explorer/ExplorerSpec.hs:
```diff
@@ -78,158 +100,247 @@ getTransactionsPaged chunkSize mAddress = getFrom Nothing
+                         else Just (skAddress sk, edCvMap)
+
+            let fairCv = FairCV . M.fromList $ catMaybes cvPairs
+            fairCvInvalid <- spoilFairCV fairCv
```
:yao:
But yeah, worth adding `suchThat (not . null)` above.

Martoon-00 replied:
How did you think we came up with periodically failing CI in master before :iswydt:

flyingleafe commented on witness/tests/Test/Dscp/Witness/Explorer/ExplorerSpec.hs:
```diff
@@ -78,158 +100,247 @@ getTransactionsPaged chunkSize mAddress = getFrom Nothing
+        oldAddr <- pick . elements $ M.keys cv
+        newAddr <- pick arbitrary
+        let subCv = fromMaybe (error "impossible") $
+                    M.lookup oldAddr cv
```
@Martoon-00 Which pair? Not sure I got what you mean.

Martoon-00 replied:
```
(key, value) <- elements (M.toList cv)
```
instead of
```
key <- elements (M.keys) cv
let value = fromMaybe (error "impossible") $
                 M.lookup oldAddr cv 
```
There seems to be no performance overhead of the former comparing to the latter, `elements` converts map to the list inside anyway.

flyingleafe commented on witness/src/Dscp/Witness/Web/Types.hs:
```diff
@@ -152,6 +152,9 @@ instance Buildable (ForResponseLog HashIs) where
 
+instance Buildable (ForResponseLog FairCVCheckResult) where
+    build (ForResponseLog res) = build res
+
```
@Martoon-00 Since the responses are just bool (until DSCP-404), let's just print the number of failures and number of successes here. In DSCP-404, we may elaborate on that.

Martoon-00 replied:
Noice :+1:

flyingleafe committed:
[DSCP-377] Address PR comments

Martoon-00 commented on witness/src/Dscp/Witness/Logic/Getters.hs:
```diff
@@ -38,17 +39,22 @@ import Dscp.Witness.Logic.Exceptions
+getTipHash =
+    fmap (fromMaybe genesisHash . SD.unTipValue) $
+    SD.queryOne SD.TipKey >>=
+    nothingToThrowSD @(SD.BlockStateException Ids) SD.TipNotFound
```
:+1:

Martoon-00 commented on witness/tests/Test/Dscp/Witness/Explorer/ExplorerSpec.hs:
```diff
@@ -334,13 +334,24 @@ spec = describe "Explorer" $ do
+            -- If we somehow generated an empty FairCV, just skip and
+            -- have a better luck next time.
+            if null cvPairs
+                then return $ property True
```
Better use `pre`, it would make the test discarded rather than accepted (although this is minor here).

flyingleafe commented:
@Martoon-00 pls re-review

Martoon-00 commented on core/src/Dscp/Core/FairCV.hs:
```diff
@@ -0,0 +1,157 @@
+        let mergeUntag t t' =
+                unTaggedProof <$>
+                mergeProofs (TaggedProof t) (TaggedProof t')
+        in ProofBranch s <$> mergeUntag l l' <*> mergeUntag r r'
```
By the way, you do not check that the signatures on the left and on the right sides match.
I suppose it means that merging two validated proofs can produce invalid proof if this function is used inaccurately.
Would be nice to specify more concrete preconditions in haddock.

flyingleafe replied:
@Martoon-00 what do you mean? signature match is checked for _every_ pair of nodes, see line 73.

Martoon-00 replied:
Oh right, missed it this time.

Martoon-00 reviewed (approved):
LGTM
Just address one more minor comment please.
